### PR TITLE
Data hash includes b58

### PR DESCRIPTION
**Pull request by emberian**
We keep having problems where the decimal string shows up. Remove those functions entirely, in favor of having data hash use the base58_check module to define them.

This solves all the data hash inconsistencies #4810.

### CODE REVIEW

File: src/lib/coda_base/data_hash.ml [hunk 1]
```diff
@@ -146,13 +137,30 @@ struct
...
+  [%%define_locally
+  Base58_check.(to_base58_check, of_base58_check, of_base58_check_exn)]
+
+  [%%define_locally
+  Base58_check.String_ops.(to_string, of_string)]
+
+  [%%define_locally
+  Base58_check.(to_yojson, of_yojson)]
+
   (* inside functor of no arguments, versioned types are allowed *)
```
Comment (emberian) [inline src/lib/coda_base/data_hash.ml:164]: @psteckler this comment is what I'm now violating - should this be structured differently? Do you have some advice?

Comment (psteckler) [reply src/lib/coda_base/data_hash.ml:164]: @emberian The solution we've used is to have an unversioned type inside the functor, and create a versioned type at each place the functor is applied. 

An example is in `Consensus.Proof_of_stake`, the module `Staking_value_versioned`.

**Comment by emberian** [comment]
@psteckler this ended up being way more painful than I anticipated... but here it is! It is now impossible to define a data hash without defining also the base58_check stuff correctly. What do you think to the changes in data_hash.ml? The rest was pretty much boilerplate.

**Comment by emberian** [comment]
Oh, I also took the liberty of renumbering all the version bytes to be sequential starting with 1, except for private key (so as to not break any extant encodings in the testnet).

File: src/lib/base58_check/version_bytes.ml [hunk 2]
```diff
@@ -4,48 +4,54 @@ type t = char
 
 (* each of the following values should be distinct *)
```
Comment (Schmavery) [inline src/lib/base58_check/version_bytes.ml:5]: I'm all for having these be a little less random, but might this version byte change force us to do the whole "generate a new public key" thing next release?

Comment (emberian) [reply src/lib/base58_check/version_bytes.ml:5]: Hm, yes, and it's not the private_key byte that needs to stay the same, the non_zero_curve_points are the actually-important ones.

Comment (emberian) [reply src/lib/base58_check/version_bytes.ml:5]: I think I've preserved the important values now.

File: src/lib/base58_check/version_bytes.ml [hunk 3]
```diff
@@ -4,48 +4,58 @@ type t = char
...
+let random_oracle_base : t = '\x0B'
 
-let private_key : t = '\x5A'
+let receipt_chain_hash : t = '\x0C'
 
-let proof : t = '\x70'
+let epoch_seed : t = '\x0D'
 
-let random_oracle_base : t = '\x03'
+let staged_ledger_hash_aux_hash : t = '\x0F'
```
Comment (psteckler) [inline src/lib/base58_check/version_bytes.ml:27]: skip \x0E ?

File: src/lib/coda_base/ledger_hash.ml [hunk 4]
```diff
@@ -34,24 +34,37 @@ module Merkle_tree =
...
+  module V1 = struct
+    module T = struct
+      type t = Field.t [@@deriving sexp, compare, hash, version {asserted}]
+    end
+
+    include T
+
+    let to_latest = Core.Fn.id
+
+    let to_yojson = to_yojson
```
Comment (psteckler) [inline src/lib/coda_base/ledger_hash.ml:56]: think you can use the newfangled `[%%define_from_scope (to_yojson,of_yojson)]`

File: src/lib/coda_base/pending_coinbase.ml [hunk 5]
```diff
@@ -150,16 +150,38 @@ module type Data_hash_binable_intf = sig
...
+    module V1 = struct
+      module T = struct
+        type t = Field.t [@@deriving sexp, compare, hash, version {asserted}]
+      end
+
+      include T
+
+      let to_latest = Core.Fn.id
+
+      let to_yojson = to_yojson
```
Comment (psteckler) [inline src/lib/coda_base/pending_coinbase.ml:175]: ditto re define_from_scope

File: src/lib/coda_base/pending_coinbase.ml [hunk 6]
```diff
@@ -189,7 +211,35 @@ module Coinbase_stack_data = struct
...
```
Comment (psteckler) [inline src/lib/coda_base/pending_coinbase.ml:233]: ditto

File: src/lib/coda_base/ledger_hash.ml (hunk 4 above)
Comment (psteckler) [reply src/lib/coda_base/ledger_hash.ml:56]: or not, more useful when the list is long.

Comment (emberian) [reply src/lib/coda_base/ledger_hash.ml:56]: aha, I was looking at define_locally and saw a TODO about it. glad to know this exists!

**Pull request merged**